core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/leonardo/LeonardoService.scala

```diff
@@ -29,110 +31,91 @@ class LeonardoService(leonardoDAO: LeonardoDAO)(implicit
         }
       }
     }
 
-  def deleteApps(workspace: Workspace, ctx: RawlsRequestContext)(implicit ec: ExecutionContext): Future[Unit] =
-    retry(when500OrProcessingException) { () =>
-      Future {
-        blocking {
-          logger.info(s"Sending app deletion request [workspaceId=${workspace.workspaceIdAsUUID}]")
-          leonardoDAO.deleteApps(ctx.userInfo.accessToken.token, workspace.workspaceIdAsUUID, deleteDisk = true)
-        }
-      }
+  def listRunningRuntimes(workspace: Workspace, ctx: RawlsRequestContext)(implicit
+    ec: ExecutionContext
+  ): Future[Seq[ListRuntimeResponse]] =
+    getAllRuntimes(workspace, ctx).map { allRuntimes =>
+      val statuses = Set(ClusterStatus.RUNNING,
+                         ClusterStatus.STARTING,
+                         ClusterStatus.STOPPING,
+                         ClusterStatus.CREATING,
+                         ClusterStatus.UPDATING,
+                         ClusterStatus.DELETING
+      );
+      allRuntimes.filter(runtime => statuses.contains(runtime.getStatus));
     }
 
-  def deleteRuntimes(workspace: Workspace, ctx: RawlsRequestContext)(implicit ec: ExecutionContext): Future[Unit] =
+  private def getAllDisks(workspace: Workspace, ctx: RawlsRequestContext)(implicit
+    ec: ExecutionContext
+  ): Future[Seq[ListPersistentDiskResponse]] =
     retry(when500OrProcessingException) { () =>
       Future {
         blocking {
-          logger.info(s"Sending runtime deletion request [workspaceId=${workspace.workspaceIdAsUUID}]")
-          leonardoDAO.deleteAzureRuntimes(ctx.userInfo.accessToken.token,
-                                          workspace.workspaceIdAsUUID,
-                                          deleteDisk = true
-          )
+          val allDisks = leonardoDAO.listDisks(ctx.userInfo.accessToken.token, workspace.googleProjectId);
+          allDisks.filter { disk =>
+            val cloudContext = disk.getCloudContext
+            cloudContext != null &&
+            cloudContext.getCloudResource == workspace.googleProjectId.value &&
+            cloudContext.getCloudProvider == CloudProvider.GCP
+          }
         }
       }
     }
 
+  def listRunningDisks(workspace: Workspace, ctx: RawlsRequestContext)(implicit
+    ec: ExecutionContext
+  ): Future[Seq[ListPersistentDiskResponse]] =
+    getAllDisks(workspace, ctx).map { allDisks =>
+      val statuses = Set(DiskStatus.CREATING, DiskStatus.READY, DiskStatus.RESTORING, DiskStatus.DELETING);
+      allDisks.filter(disk => statuses.contains(disk.getStatus));
+    }
+
+  def hasActiveResources(workspace: Workspace, ctx: RawlsRequestContext)(implicit
+    ec: ExecutionContext
+  ): Future[Boolean] =
+    for {
+      runtimes <- listRunningRuntimes(workspace, ctx)
+      apps <- listRunningApps(workspace, ctx)
+      disks <- listRunningDisks(workspace, ctx)
+    } yield runtimes.nonEmpty || apps.nonEmpty || disks.nonEmpty
```

Comment:
I'm waffling a bit here ... as-is, this will always issue three requests to Leo, but it issues them in parallel.

Instead, we could call `listRunningDisks` first and inspect its result. If any disks are present, we could return and skip the other two requests. If there are no disks, we could try runtimes next, and finally only ask for apps if there are no runtimes. But these would all happen serially.

What's better - potentially extraneous API calls in parallel, or reduce API calls but force serialization?

Reply:
Sounds fair to me, giving this a :+1: